src/coap_client_zephyr.c

```diff
@@ -0,0 +1,1072 @@
+    if (golioth_sys_now_ms() > req->ageout_ms) {
+        LOG_WRN("Ignoring request that has aged out, type %d, path %s",
+                req->type,
+                (req->path ? req->path : "N/A"));
+
+        if (req->type == GOLIOTH_COAP_REQUEST_POST && req->post.payload_size > 0) {
+            golioth_sys_free(req->post.payload);
+        }
+
+        if (req->request_complete_event) {
+            golioth_event_group_destroy(req->request_complete_event);
+            golioth_sys_sem_destroy(req->request_complete_ack_sem);
+        }
+
+        goto free_req;
+    }
+
+    req->client = client;
+
+    // Handle message and send request to server
+    switch (req->type) {
+        case GOLIOTH_COAP_REQUEST_EMPTY:
+            LOG_DBG("Handle EMPTY");
+            golioth_send_coap_empty(req->client);
+            goto free_req;
+        case GOLIOTH_COAP_REQUEST_GET:
+            LOG_DBG("Handle GET %s", req->path);
+            err = golioth_coap_req_cb(req->client, COAP_METHOD_GET,
+                                      PATHV(req->path_prefix, req->path),
+                                      golioth_content_type_to_coap_format(req->get.content_type),
+                                      NULL, 0,
+                                      golioth_coap_cb, req,
+                                      0);
+            break;
+        case GOLIOTH_COAP_REQUEST_GET_BLOCK:
+            LOG_DBG("Handle GET_BLOCK %s", req->path);
+            err = golioth_coap_get_block(req);
+            break;
+        case GOLIOTH_COAP_REQUEST_POST:
+            LOG_DBG("Handle POST %s", req->path);
+            err = golioth_coap_req_cb(req->client, COAP_METHOD_POST,
+                                      PATHV(req->path_prefix, req->path),
+                                      golioth_content_type_to_coap_format(req->post.content_type),
+                                      req->post.payload, req->post.payload_size,
+                                      golioth_coap_cb, req,
+                                      0);
+            golioth_sys_free(req->post.payload);
+            break;
+        case GOLIOTH_COAP_REQUEST_DELETE:
+            LOG_DBG("Handle DELETE %s", req->path);
+            err = golioth_coap_req_cb(req->client, COAP_METHOD_DELETE,
+                                      PATHV(req->path_prefix, req->path),
+                                      COAP_CONTENT_FORMAT_APP_OCTET_STREAM /* not used */,
+                                      NULL, 0,
+                                      golioth_coap_cb, req,
+                                      0);
+            break;
+        case GOLIOTH_COAP_REQUEST_OBSERVE:
+            LOG_DBG("Handle OBSERVE %s", req->path);
+            err = golioth_coap_observe(req, client);
+            break;
+        default:
+            LOG_WRN("Unknown request_msg type: %u", req->type);
+            goto free_req;
+    }
+
+    if (err) {
+        status = GOLIOTH_ERR_MEM_ALLOC;
```

Comment:
Is this the only error that can occur?

Reply:
Good catch. I have added a function to map from `err` to `status` for most error codes I can think of. Also changed a bit the implementation, so that `status` is not ignored. See https://github.com/golioth/golioth-firmware-sdk/compare/00f2b0e96d95b498ba28e0ced956a195b03d6997..48cbd0a5dec09c5557a0c623ace897ded7bc5502